crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -890,45 +890,67 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
 
-        let oflags = OFlags::from(&oflags);
-        let fdflags = FdFlags::from(fdflags);
         let path = path.as_str()?;
-        if oflags.contains(OFlags::DIRECTORY) {
-            if oflags.contains(OFlags::CREATE)
-                || oflags.contains(OFlags::EXCLUSIVE)
-                || oflags.contains(OFlags::TRUNCATE)
-            {
-                return Err(Error::invalid_argument().context("directory oflags"));
+        let path = path.deref();
+        match dir.get_path_filestat(path, symlink_follow).await {
+            Ok(Filestat {
+                filetype: FileType::Directory,
+                ..
+            }) if non_dir_oflags => {
+                Err(Error::invalid_argument().context("oflags not applicable to a directory"))
             }
-            let dir_caps = dir_entry.child_dir_caps(DirCaps::from(&fs_rights_base));
-            let file_caps = dir_entry.child_file_caps(FileCaps::from(&fs_rights_inheriting));
-            let dir = dir_entry.get_cap(DirCaps::OPEN)?;
-            let child_dir = dir.open_dir(symlink_follow, path.deref()).await?;
-            drop(dir);
-            let fd = table.push(Box::new(DirEntry::new(
-                dir_caps, file_caps, None, child_dir,
-            )))?;
-            Ok(types::Fd::from(fd))
-        } else {
-            let mut required_caps = DirCaps::OPEN;
-            if oflags.contains(OFlags::CREATE) {
-                required_caps = required_caps | DirCaps::CREATE_FILE;
+
+            Ok(Filestat {
+                filetype: FileType::Directory,
+                ..
+            }) => {
+                let dir_caps = dir_entry.child_dir_caps(DirCaps::from(&fs_rights_base));
+                let file_caps = dir_entry.child_file_caps(FileCaps::from(&fs_rights_inheriting));
+                let child_dir = dir.open_dir(symlink_follow, path).await?;
+                drop(dir);
+                let fd = table.push(Box::new(DirEntry::new(
+                    dir_caps, file_caps, None, child_dir,
+                )))?;
+                Ok(types::Fd::from(fd))
             }
 
-            let file_caps = dir_entry.child_file_caps(FileCaps::from(&fs_rights_base));
-            let dir = dir_entry.get_cap(required_caps)?;
-            let read = file_caps.contains(FileCaps::READ);
-            let write = file_caps.contains(FileCaps::WRITE)
-                || file_caps.contains(FileCaps::ALLOCATE)
-                || file_caps.contains(FileCaps::FILESTAT_SET_SIZE);
-            let file = dir
-                .open_file(symlink_follow, path.deref(), oflags, read, write, fdflags)
-                .await?;
-            drop(dir);
-            let fd = table.push(Box::new(FileEntry::new(file_caps, file)))?;
-            Ok(types::Fd::from(fd))
+            Ok(_) if oflags.contains(OFlags::DIRECTORY) => {
+                Err(Error::not_dir().context("directory requested"))
+            }
+
```

Comment:
Ideally, I would want to fail here if error is not `not found`, but that would require either cloning the error (to be able to propagate it if it does not match) and then converting it into a `types::Errno`. Alternatively, it should be possible to downcast the error, but would require matching on all possible error types, which seems like the wrong thing to do as well.
Is there a reasonable way to do that? I'm looking for `Error::not_found(&self) -> bool`
Discard the error instead and preserve the original behavior.

Reply:
I think it is reasonable to downcast_ref and match for an ErrorKind::Noent, and if that fails downcast to a std::io::Error with the appropriate check there as well. This is a big janky, and I think if I had to redesign wasi-common I probably wouldn't use a fully dynamic error type and instead make `enum Error { Std(std::io::Error), Wasi(ErrorKind) }` or something like that, but its been a while since I had this system in my head so I'm not sure why I didn't do that in the first place.